girder/api/v1/resource.py

```diff
@@ -207,6 +210,51 @@ def stream():
             yield zip.footer()
         return stream
 
+    @access.public(scope=TokenScope.DATA_READ)
+    @autoDescribeRoute(
+        Description('Download a resource based on its path')
+        .param('path', 'The path of the resource.', paramType='path')
+        .errorResponse('Path is invalid.')
+        .errorResponse('Path refers to a resource that does not exist.')
+        .errorResponse('Read access was denied for the resource.', 403)
+    )
+    def pathDownload(self, path):
+        user = self.getCurrentUser()
+        path = '/'.join(path)
+        resource = path_util.lookUpPath(path, self.getCurrentUser())['document']
+        if resource['_modelType'] == 'file':
+            single_file = resource
+        else:
+            model = self._getResourceModel(resource['_modelType'], 'fileList')
+            single_file = None
+            for (path, file) in model.fileList(doc=resource, user=user, subpath=True, data=False):
+                if single_file is None:
+                    single_file = file
+                else:
+                    single_file = False
+                    break
+        if single_file is None:
```

Comment:
So `single_file` could be either `None`, or `False`, or a `dict`, with significantly different behavior for each. This might be abusing dynamic typing a little much, but that's just my opinion.

Reply:
Why is the use of a language feature abuse?  The alternative is to use multiple variables, which, in my opinion, reduces readability rather than increases it.